modules_for_BSAT/protein_analysis.py

```diff
@@ -0,0 +1,316 @@
+
+# amino acid with corresponding degenerate codon/codons
+AMINO_ACID_TO_MRNA = {'A': 'GCN',
+                      'R': '(CGN/AGR)',
+                      'N': 'AAY',
+                      'D': 'GAY',
+                      'C': 'UGY',
+                      'Q': 'CAR',
+                      'E': 'GAR',
+                      'G': 'GGN',
+                      'H': 'CAY',
+                      'I': 'AUH',
+                      'L': '(CUN/UUR)',
+                      'K': 'AAR',
+                      'M': 'AUG',
+                      'F': 'UUY',
+                      'P': 'CCN',
+                      'S': '(UCN/AGY)',
+                      'T': 'ACN',
+                      'W': 'UGG',
+                      'Y': 'UAY',
+                      'V': 'GUN'}
+
+def save_register(seq: str) -> List[int]:
+    """
+    Additional function for correct change_residues_encoding return
+    :param seq: protein seq (str)
+    :return: List of type of register character in protein seq (List[num])    
+    """
+    register = []
+    sep_seq = []
+    if ' ' in seq:
+        sep_seq = seq.split()
+    else:
+        sep_seq = list(seq)
+    for residue in sep_seq:
+        if residue.isupper():
+            register.append(1)
+        else:
+            register.append(0)
+    return register
+
+
+def change_residues_encoding(seq: str, encoding = 'one') -> str:
+    """
+    Transfer amino acids from 3-letter to 1-letter code. By default, converts all seq into 1-letter
+    :param seq: protein seq (str)
+    :param encoding: specify target encoding (str)
+    :return: same protein seq in another encoding (str)
+    """
+    encode_seq = []
+    encode_seq_registered = []
+    if ' ' in seq:
+        sep_seq = seq.upper().split()
+    else:
+        sep_seq = list(seq)
+    residue_code_length = len(sep_seq[0])
+    for residue in sep_seq:
+        if len(residue) != residue_code_length:
+            raise ValueError (f'Wrong sequence format in {seq}!')
+        if encoding == 'one':
+            if len(sep_seq[0]) == 1:
+                encode_seq.append(residue)
+            if len(sep_seq[0]) == 3:
+                encode_seq.append(RESIDUES_NAMES[residue])
+        if encoding == 'three':
+            if len(sep_seq[0]) == 3:
+                encode_seq.append(residue)
+            if len(sep_seq[0]) == 1:
+                encode_seq.append(RESIDUES_NAMES_THREE[residue])
+    for residue, reg in zip(encode_seq, save_register(seq)):
+        if (reg == 1):
+            encode_seq_registered += residue.upper()
+        elif (reg == 0):
+            encode_seq_registered += residue.lower()
+    if encoding == 'one':
+        return ''.join(encode_seq_registered)
+    if encoding == 'three':
+        fin_seq = []
+        for i, residue in enumerate(encode_seq_registered):
+            fin_seq.append(residue)
+            if ((i+1) % 3 == 0):
+                fin_seq.append(' ')
+        return ' '.join(fin_seq)
```

Comment:
```suggestion
        return ''.join(fin_seq)
```
Лучше убрать пробел, а то запись странная получается. 
<img width="997" alt="Снимок экрана 2023-10-16 в 12 02 14" src="https://github.com/grishchenkoira/BSAT/assets/90711855/cea24441-7f76-4b15-b162-ed70333e0c97">


Код работает правильно, но в нем много повторений. Если тебе будет интересно, как это можно сделать короче, напиши мне, я напишу альтернативную версию.